Перенос состояния в Context

EldenLordCommander opened the pull request:
Добрый день.
Как я понял из задачи, для данной промежуточной работы нужны в основном подготовительные работы:
- добавил работу с Context вместо пропс для Ингредиентов и Конструктора, 
- добавил вызов Апи для номера заказа для модалки, 
- добавил заполнения массива Ингредиентов из конструктора по клику на них, для дальнейшей отправки для получения номера заказа
- добавил функции вычисления стоимости: начальная сумма равна стоимости двух булок (верх и низ), увеличение стоимости происходит по клику на ингредиент. Вычитание работает аналогично, если поменять тип в вызове
- в самом конструкторе по поводу булок, как я понял, пока добавить нечего, так как у меня итак булки только сверху и снизу, в основном списке булок нет, цену их для стоимости я высчитываю при рендере.

Буду рад, если поправите меня, если я не прав или не так понял

EldenLordCommander committed:
Перенос состояния в Context

Добавил работу с Context вместо пропс для Ингредиентов и Конструктора, добавил вызов Апи для номера заказа для модалки, добавил функции вычисления стоимости

Zaytsev-Alex commented on src/components/app/app.tsx:
```diff
@@ -1,44 +1,48 @@
    useEffect(() => {
       fetch(url)
          .then((response) => {
             if (!response.ok) {
```
Отлично, что есть обработка ответа сервера. Так корректно обработаются ответы с статусом отличным от 200-ого.

Zaytsev-Alex commented on src/components/app/app.tsx:
```diff
@@ -1,44 +1,48 @@
    const url = "https://norma.nomoreparties.space/api/ingredients";
 
    useEffect(() => {
       fetch(url)
```
Можно лучше: функции для работы с API стоит вынести в отдельный файл, к примеру, utils/burger-api.js. Так их код поддерживать будет легче, т.к. не надо будет искать их в компонентах. Также компоненту не за чем знать, как реализована эта функция, и наоборот, функции не стоит иметь в своей видимости данные компонента.
Дополнительно удобно будет вынести проверку ответа res.ok в функцию, чтобы переиспользовать ее.

Zaytsev-Alex commented on src/components/app/app.tsx:
```diff
@@ -1,44 +1,48 @@
+      data: [],
+      success: false
    });
    const url = "https://norma.nomoreparties.space/api/ingredients";
```
Надо исправить: базовый адрес сервера необходимо вынести в отдельную константу в отдельном файле, к примеру, том же utils/burger-api.js. Далее адреса эндпоинтов создавать используя ее. Базовый адрес сервера меняется сравнительно часто, может быть адрес для тестирования, для разработки, для продакшена и тд. А адрес эндпоинта обычно как был /ingredients, так и остается. Код может быть примерно такой:
```
const BASE_API_URL = 'https://norma.nomoreparties.space/api';
export const INGREDIENTS_URL = `${BASE_API_URL}/ingredients`;
export const ORDERS_URL = `${BASE_API_URL}/orders`;
```

Zaytsev-Alex commented on src/components/app/app.tsx:
```diff
@@ -1,44 +1,48 @@
 import BurgerConstructor from '../burger-constructor/burger-constructor';
 //import Data from '../../utils/data.js'
 
+export const IngredientContext = React.createContext([]);
```
Надо исправить: создание контекста стоит вынести в отдельный файл, services/burger-context.js, чтобы была возможность его переиспользовать между компонентов. Не стоит его экспортировать из файла с компонентом, файл с компонентом должен отвечать только за компонент.

Zaytsev-Alex commented on src/components/app/app.tsx:
```diff
@@ -1,44 +1,48 @@
-
-               <BurgerIngredients data={state.data} />
-               <BurgerConstructor data={state.data} />
+               <IngredientContext.Provider value={state.data}>
```
Отлично, что в провайдер контекста обернуты только нужные компоненты.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,22 +1,85 @@
     const firstIngredient = data[0];
     const leftIngredients = data.slice(1, data.length - 1);
     const [openModal, setModal] = useState(false);
+    const url = 'https://norma.nomoreparties.space/api/orders';
```
Надо исправить: замечания аналогичные компоненту App.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,22 +1,85 @@
     const orderClick = () => {
+        getOrderId();
         setModal(true);
+        console.log(clickedIngredients);
```
Можно лучше: ненужные консоль логи из проекта стоит удалять. Можно оставлять консоль логи, логирующие какие-то ошибки, чтобы приложение даже на стадии продакшена можно было поддерживать.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,22 +1,85 @@
 
-    return (
+    const getOrderId = () => {
+        fetch(url, {
```
Можно лучше: аналогично, стоит вынести функцию в отдельный файл.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,22 +1,85 @@
+            }, 
+            body: JSON.stringify({
+                //ingredients : data.map((ingredient)=>ingredient._id)})
+                ingredients : clickedIngredients })
```
Надо исправить: в заказ должны помещаться две булочки: верхняя и нижняя, также они должны стоять на правильных местах: в начале и конце массива ингредиентов.

EldenLordCommander replied:
Добрый вечер @Zaytsev-Alex : вам сюда можно вопрос задать, или лучше в один из чатов Слака?

По поводу этого исправления: 
**> Добавлять в заказ две булочки на нужных местах;**
 имеется ввиду, что в массиве, которые передается в АПИ для получения номера заказа ПЕРВЫЙ и ПОСЛЕДНИЙ элемент должен быть id булок Верх и Низ? А содержимое между ними можно будет менять, правильно?

Zaytsev-Alex replied:
Все правильно!

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,22 +1,85 @@
+           .then(
+              result => setOrder({ name: result.name, order: result.order, success: result.success })
+           )
+           .catch(error => alert("Ошибка запроса: " + error))
```
Отлично, что в случае ошибки пользователь будет уведомлен!

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,22 +1,85 @@
+
+    const initialPriceState = { price: firstIngredient.price * 2 };
+    //const [price, setPrice] = useState(null);
+    const [priceState, priceDispatcher] = useReducer(reducer, initialPriceState);
```
Отлично, что вы получили опыт работы с useReducer, это полезный хук, который позволяет делать более гибкие и комплексные изменения данных по сравнению с useState.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -84,8 +148,13 @@ function BurgerConstructor({ data }) {
+// }
 BurgerConstructor.propTypes = {
-    data: PropTypes.arrayOf(ingredientPropTypes).isRequired,
+    IngredientContext: PropTypes.shape({
+        data: PropTypes.arrayOf(ingredientPropTypes).isRequired,
+    })
```
Можно лучше: нет необходимости в типизации контекста.

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -98,8 +102,13 @@ function BurgerIngredients({ data }) {
+// }
 BurgerIngredients.propTypes = {
-    data: PropTypes.arrayOf(ingredientPropTypes).isRequired,
+    IngredientContext: PropTypes.shape({
+        data: PropTypes.arrayOf(ingredientPropTypes).isRequired,
+    })
```
Можно лучше: нет необходимости в типизации контекста.

Zaytsev-Alex commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -98,8 +102,13 @@ function BurgerIngredients({ data }) {
     )
 }
 
+// BurgerIngredients.propTypes = {
+//     data: PropTypes.arrayOf(ingredientPropTypes).isRequired,
+// }
```
Можно лучше: закомментированный код из проекта стоит удалять, т.к. он его загрязняет и усложняет поддержку. При необходимости этот код можно найти используя git.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,22 +1,85 @@
 
     const orderClick = () => {
+        getOrderId();
         setModal(true);
```
Надо исправить: модальное окно  с деталями заказа должно отображаться только в случае успешного создания. Стоит проверять так же статус заказа, проверяя поле success.

Zaytsev-Alex reviewed (commented):
Отличная работа, весь требуемый функционал реализован. Правильно организована работа с контекстом, круто, что был использован useReducer. Также хочу отметить, замечательно ведение истории: коммиты имеют подробное описание, PR также с описанием!

Из того, что надо исправить:
- Вынести базовый адрес сервера в отдельную константу в отдельном файле;
- Отображать модальное окно с деталями заказа только при успешном создании заказа;
- Добавлять в заказ две булочки на нужных местах;
- Вынести создание контекст в отдельный файл.

Из того, что можно сделать лучше, хочу отметить следующее:
- Вынести функции для работы с API в отдельный файл;
- Убирать из проекта ненужные консоль логи и закомментированный код;
- Нет необходимости в типизации контекста, используя prop-types.

Более детальные комментарии можно найти на протяжении проверки.

Хочу напомнить, что пункты помеченные как "надо исправить" будет обязательны к исправлению в следующем задании, но можно их исправить и сейчас :)

Хорошего вечера!

EldenLordCommander committed:
Исправление недочетов по ревью

Вынос url и запросов в отдельные файлы, удаление console.log, закомментированного кода, правки в вызове модалки и формировании массива заказа

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -16,70 +17,55 @@ function BurgerConstructor() {
-        setModal(true);
-        console.log(clickedIngredients);
-        
+        getOrderId([firstIngredient._id, ...clickedIngredients, firstIngredient._id]).then((result) => {
```
Отлично, теперь правильно. Касательно вашего вопроса, все верно. первый и последний ингредиент в заказе - булочка, между ними - начинки, добавленные в бургер.

Zaytsev-Alex commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -16,70 +17,55 @@ function BurgerConstructor() {
+            setOrder({ name: result.name, order: result.order, success: result.success });
+            if(result.success)
+            {
+                setModal(true);
```
Можно лучше: по сути можно обойтись только номером заказа в стейте. Т.к. название заказа не используется, в стейт можно помещать только его номер, если success заказа правдивый. Это позволит избавиться от лишнего стейта, в том числе [openModal, setModal]. Если номер заказа установлен - модальное окно с деталями заказа отображается, иначе нет.

Zaytsev-Alex commented on src/utils/burger-api.js:
```diff
@@ -0,0 +1,47 @@
+export const getIngredients = () => {
+    return fetch(INGREDIENTS_URL)
+        .then((response) => {
+            if (!response.ok) {
+                throw new Error('Ответ от сервера вернул ошибку');
+            }
+            return response.json();
```
Можно лучше: проверка ответа сервера одинаковая для этих функций. Можно создать функцию checkResponse, которая будет делать:
```
if (!response.ok) {
      throw new Error('Ответ от сервера вернул ошибку');
}
return response.json();
```
и использовать ее, чтобы избавиться от дублирования кода.

Zaytsev-Alex commented on src/utils/burger-api.js:
```diff
@@ -0,0 +1,47 @@
+            }
+        }
+        )
+        .catch(error => alert("Ошибка запроса: " + error))
```
Можно лучше: из этих функций стоит возвращать Promise.reject в случае ошибки и обрабатывать саму ошибку (ее вывод в alert) в компоненте. Функции, работающие с API, должны работать только с API, сделать запрос и обработать ответ, далее все действия делает компонент. Таким образом функции получатся более гибкими для переиспользования.

Zaytsev-Alex reviewed (approved):
Отлично, все критические замечания исправлены, работа принимается, поздравляю!
Радует, что были реализованы и советы по улучшению кода.

Из того, что можно сделать лучше, по результатам этой итерации, могу выделить следующее:
- Использовать только номер заказа в качестве стейта и удалить оставшиеся данные относящиеся к заказу из стейта;
- Вынести проверку ответа res.ok в отдельную функцию;
- Отображать ошибку пользователю при запросе в компонента, а не в функции, делающей запрос.

Желаю удачи в дальнейшем обучении!
